### PR TITLE
Extend Mm tag to allow callers to indicate unknown positions

### PR DESCRIPTION
**Pull request by jts**
This draft PR addresses the issue with the `Mm` tag raised in #582. It adds an optional `.` or `?` to change how downstream tools should interpret the skipped bases. When `?` is present, nothing should be assumed about the modification status of the skipped bases. This proposal is slightly different than what we originally discussed in #582:

- the flag is now optional as suggested by @jkbonfield
- the flag now comes after the modification code, as I found this to be nicer when optional
- I added an example of how `?` should be interpreted
- I added a footnote to guide interpretation of the (lack of) modification probabilities for skipped bases in `.` mode

Also, I reworded some existing text to reflect the uncertainty present in either mode. In some places the existing text implied that any bases not skipped are modified, but this needs to be determined by looking at the `Ml` tag to see if the probability is > 0.5. So I added words like "potentially modified" in a number of places.

### CODE REVIEW

**Comment by jts** [comment]
Hi @jkbonfield, could I gently bump this? I'd like to release a new version of nanopolish with modbam support, and would like to see this merged in (if acceptable) before I make that release.

**Comment by jkbonfield** [comment]
My apologies Jared.  I've been concentrating my efforts elsewhere and this has slipped.  I'll take a look.

File: SAMtags.tex [hunk 1]
```diff
@@ -498,26 +498,33 @@ \subsection{Base modifications}
...
 albeit counting specific base types only and potentially reverse-complemented.
 
 For example {\tt C+m,5,12,0;} tells us there are three
-5-Methylcytosine bases on the top strand of {\sf SEQ}.
-The first 5 {\tt C} bases are unmodified and the 6th is modified, as are the 19th (with 12 between the 6th and 19th) and 20th.
-Similarly {\tt G-m,14;} indicates the 15th {\tt G} is a 5-Methylcytosine on the opposite strand (still counting using the top strand base calls from the 5' end).
+potential 5-Methylcytosine bases on the top strand of {\sf SEQ}.
+The first 5 {\tt C} bases are unmodified and the 6th, 19th and 20th have modification status indicated by the corresponding probabilities in the {\tt Ml} tag. The 12 cytosines between the 6th and 19th cytosine are unmodified. Modification probabilities for the 17 skipped cytosines are not provided.
+
+When the {\tt ?} flag is present the tag {\tt C+.m?,5,12,0;} tells us the modification status of the first five
```
Comment (jkbonfield) [inline SAMtags.tex:524]: `C+.m?,5,12,0;` has both `.` and `?` present, in two different locations.  I don't understand this syntax.  Did you mean `C+?m,5,12,0;`?

File: SAMtags.tex [hunk 2]
```diff
@@ -498,26 +498,33 @@ \subsection{Base modifications}
 This potentially differs to the sequence stored in the main SAM {\sf SEQ} field if the latter has been reverse complemented, in which case SAM {\sf FLAG} 0x10 must be set.
 This means modification positions are also recorded against the original orientation (i.e. starting at the 5' end), and count the original base types.
 
-Each modified base listed also has a quality value associated with it.
+Each modified base prediction listed also has a quality value associated with it.
 Given the unmodified base already has a phred likelihood, this base modification quality should be interpreted as the likelihood of this modification being correct given an assumption the original call is correct.
 
 \begin{description}
-\item[Mm:Z:\tagregex{([ACGTUN][-+]([a-z]+|[0-9]+)(,[0-9]+)*;)*}]
+\item[Mm:Z:\tagregex{([ACGTUN][-+]([a-z]+|[0-9]+)(,[0-9]+)*[.?]?;)*}]
```
Comment (jkbonfield) [inline SAMtags.tex:505]: This regexp is wrong.

Breaking it down, the original was canonical_base `[ACGTUN]`, strand `[+-]`, a list of modified base types or CHEBI ID `([a-z]+|[0-9]+)`, followed by a comma-separated series of deltas: `(,[0-9]+)*` ending in a semicolon.

As far as I understand the intention, the `.` or `?` appear between the strand and base type list, so we'd want:

`([ACGTUN][-+]([a-z]+|[0-9]+)[.?]?(,[0-9]+)*;)*`

File: SAMtags.tex [hunk 3]
```diff
@@ -530,10 +537,10 @@ \subsection{Base modifications}
 When multiple modifications are listed, for example {\tt C+mh,5,12,0;}, it indicates the modification may be any of the stated bases.
 The associated confidence values in the {\tt Ml} tag may be used to determine the relative likelihoods between the options.
 The example above is equivalent to {\tt C+m,5,12,0;C+h,5,12,0;}, although this will have a different ordering of confidence values in {\tt Ml}.
-Note ChEBI codes cannot be used in the multi-modification form (such as the {\tt C+mh} example above).
+Note ChEBI codes cannot be used in the multi-modification form (such as the {\tt C+.mh} example above).
```
Comment (jkbonfield) [inline SAMtags.tex:540]: Is there a `C+.mh` example above?  We have (if fixed) a `C+?mh` one though.

**Review by jkbonfield — CHANGES_REQUESTED** [review]
Agreed with the principle, but there a number of errors in the descriptions to fix.

**Comment by jkbonfield** [comment]
> Thanks James! I'm teaching a workshop today but will address these soon.

No problem.  I'm in no position to comment on slow responses. :)

File: SAMtags.tex (hunk 1 above)
Comment (jts) [reply SAMtags.tex:524]: Sorry, the `.` were leftover from my first pass, which had a mandatory flag. I'll fix these to be `C+m?`.

File: SAMtags.tex (hunk 2 above)
Comment (jts) [reply SAMtags.tex:505]: Thanks for catching this. My intention is for the flag to appear after the base modification code - which I think is what your  proposed regex has? My regex is quite rusty...

File: SAMtags.tex (hunk 3 above)
Comment (jts) [reply SAMtags.tex:540]: I'll fix this to `C+mh`, the `.` shouldn't have been there.

File: SAMtags.tex [hunk 4]
```diff
@@ -498,26 +498,33 @@ \subsection{Base modifications}
...
 
 For example {\tt C+m,5,12,0;} tells us there are three
-5-Methylcytosine bases on the top strand of {\sf SEQ}.
-The first 5 {\tt C} bases are unmodified and the 6th is modified, as are the 19th (with 12 between the 6th and 19th) and 20th.
-Similarly {\tt G-m,14;} indicates the 15th {\tt G} is a 5-Methylcytosine on the opposite strand (still counting using the top strand base calls from the 5' end).
+potential 5-Methylcytosine bases on the top strand of {\sf SEQ}.
+The first 5 {\tt C} bases are unmodified and the 6th, 19th and 20th have modification status indicated by the corresponding probabilities in the {\tt Ml} tag. The 12 cytosines between the 6th and 19th cytosine are unmodified. Modification probabilities for the 17 skipped cytosines are not provided.
+
+When the {\tt ?} flag is present the tag {\tt C+.m?,5,12,0;} tells us the modification status of the first five 
+cytosine bases is unknown, the sixth cytosine is called, followed by 12 more unknown cytosines, and the 19th and 20th are called.
```
[the remark this answers is not in the document]
Comment (jts) [reply SAMtags.tex:525]: Good catch, I'll take your suggestion to add "(as either modified or unmodified)"

File: SAMtags.tex [hunk 5]
```diff
@@ -498,26 +498,33 @@ \subsection{Base modifications}
...
 This may be used in situations where the base modification is not a derivation of a standard base type.
-This is followed by either plus or minus indicating the strand the modification was observed on (relative to the original sequenced strand of {\sf SEQ} with plus meaning same orientation),\footnote{Hence a tool that may reverse complement sequences does not need to understand how to manipulate the {\tt Mm} and {\tt Ml} tags.} and one or more base modification codes.
-This is then followed by a comma separated list of how many unmodified seq bases of the stated base type to skip, stored as a delta to the last and starting with 0 as the first (or next) base, starting from the uncomplemented 5' end of the {\sf SEQ} field.
+This is followed by either plus or minus indicating the strand the modification was analysed for (relative to the original sequenced strand of {\sf SEQ} with plus meaning same orientation),\footnote{Hence a tool that may reverse complement sequences does not need to understand how to manipulate the {\tt Mm} and {\tt Ml} tags.} and one or more base modification codes.
```
[the remark this answers is not in the document]
Comment (jts) [reply SAMtags.tex:512]: I'll revert this back to "observed on". I changed this as "the modification was observed on" felt too definitive to me but I think it is clear enough from that context that it is OK

File: SAMtags.tex [hunk 6]
```diff
@@ -498,26 +498,33 @@ \subsection{Base modifications}
...
-This is then followed by a comma separated list of how many unmodified seq bases of the stated base type to skip, stored as a delta to the last and starting with 0 as the first (or next) base, starting from the uncomplemented 5' end of the {\sf SEQ} field.
+Following the base modification codes is an optional {\tt .} or {\tt ?} describing how skipped seq bases of the stated base type should be interpreted by downstream tools. 
+When this flag is {\tt ?} there is no information about the modification status of the skipped bases provided.
+When this flag is not present, or it is {\tt .}, these bases should be assumed to have low probability of modification\footnote{The decision whether a base is assumed to be unmodified or has a probability explicitly provided is up to the modification calling program. Some programs will elide calls with modification probabilites below a threshold to provide a more compact modification tag.}.
```
Comment (jmarshall) [inline SAMtags.tex:515]: The convention in this document is for footnotes to go after the punctuation.

File: SAMtags.tex [hunk 7]
```diff
@@ -498,26 +498,33 @@ \subsection{Base modifications}
...
 It must be one of {\tt A}, {\tt C}, {\tt G}, {\tt T}, {\tt U} (if RNA) or {\tt N} for anything else, including any IUPAC ambiguity codes in the reported SEQ field.
 Note {\tt N} may be used to match any base rather than specifically an {\tt N} call by the sequencing instrument.
 This may be used in situations where the base modification is not a derivation of a standard base type.
 This is followed by either plus or minus indicating the strand the modification was observed on (relative to the original sequenced strand of {\sf SEQ} with plus meaning same orientation),\footnote{Hence a tool that may reverse complement sequences does not need to understand how to manipulate the {\tt Mm} and {\tt Ml} tags.} and one or more base modification codes.
-This is then followed by a comma separated list of how many unmodified seq bases of the stated base type to skip, stored as a delta to the last and starting with 0 as the first (or next) base, starting from the uncomplemented 5' end of the {\sf SEQ} field.
+Following the base modification codes is an optional {\tt .} or {\tt ?} describing how skipped seq bases of the stated base type should be interpreted by downstream tools.
```
Comment (jmarshall) [inline SAMtags.tex:513]: Quote marks around the literal characters, as so (and similarly in nearby lines):

```suggestion
Following the base modification codes is an optional `{\tt .}' or `{\tt ?}' describing how skipped seq bases of the stated base type should be interpreted by downstream tools. 
```

**Review by jmarshall — APPROVED** [review]
I'm happy with the proposed notation, and pleased that the existing syntax without any `.?` flag character remains accepted.

I've pointed out a couple of formatting issues. There are a few more, but as some of those are already present in the existing text I'm happy to merge this as is (ideally with the suggested formatting fixes) and address those other concerns myself separately.

**Comment by jkbonfield** [comment]
Thanks for the review John.

I'm happy with the suggested formatting changes, so if you wish to edit and merge yourself please do so.  Or if you want me to make the changes let me know and I'll do so (and merge).

**Comment by jkbonfield** [comment]
We did also say we were going to make them official, so `mM` to `MM` etc.  We gave warning on this (https://github.com/samtools/hts-specs/pull/418#issuecomment-863365043 - long overdue), and this (welcomed) tweak was one reason for delay.  I think once this is merged we should also make the switch to official name space too, in its own commit.

**Comment by jkbonfield** [comment]
Rebased, squashed and merged as 71241fe44cabed93cb6b4c57967269270e83f78b, with John's footnote fix and quoting tidyups (and a few more nearby).

**Comment by jmarshall** [comment]
Pushed 6835cd374bc2b6a0c6ac88aa2faff4880a5edbc0 with a title page recording a commit that exists on **master**. PRs don't need to include updated PDFs (this will be done separately when they are merged), and draft PDFs containing commit hashes that exist only pre-rebasing should be undone during merging… 😄

**Pull request closed without merging**